### PR TITLE
Initial file structure changes for Package

### LINKED ISSUE #13: Make the library a proper python package
Since the package is not structured with a `setup.py` it is hard to install and consume. We would like to be able to install it like so:
```
python -m pip install git+https://github.com/VowpalWabbit/estimators.git
```

### PR DESCRIPTION
**Pull request by JuiP**
Fixes #13 @jackgerrits 
Now, it can be installed as: `python3 -m pip install vw-estimators==1.0.0`
Do let me know if any changes are needed. Thanks!

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+MINOR               = 0
+MICRO               = 0
+VERSION             = f"{MAJOR}.{MINOR}.{MICRO}"
+
+with open("README.md", "r") as f:
+    long_description = f.read()
+
+setuptools.setup(
+    name="vw-estimators",
+    version=VERSION,
```
Comment (JuiP) [inline setup.py:13]: An easy workaround to pass build (3.5) would be to include the version here. Should I make the change and `fixup`? I am a new contributor and still trying to understand the coding standards followed... :smiley:

**Comment by jackgerrits** [comment]
Hi @JuiP, 

Thanks for taking a look at the issue, I appreciate the work on the codebase. However, this issue was just about implementing the `setup.py` script and not actually releasing the code to PyPi. More work is certainly needed before this repo is ready for release. Especially a 1.0.0 release (we would be in 0.x prerelease for some time). Would it be okay if you could remove the `vw-estimators` package off of PyPi?

**Comment by JuiP** [comment]
@jackgerrits

Apologies for misunderstanding the issue. The package has been removed off of PyPi and the version number in the setup file has been updated to `"0.0.1"` in the commit cadf3c2.

I see that #10 has major additions and has been open for quite some time now.
Also just to confirm, this library will be a python implementation of the estimators already implemented in C++? Is this a natural continuation to [this project](https://vowpalwabbit.org/rlos/2021/projects#7-contextual-bandits-estimators-in-python)? More specifically my question being if IPS is already implemented in Python what is meant by the project goal?

Thanks!

**Comment by ataymano** [comment]
Hi @JuiP ,
Looks like right now all code in this repo can roughly be splitted into several categories:
- esimators themselves (i.e. ips, mle, ...). All of them seem to be implemented as contextual bandits ones, but at the same time we are going to introduce estimators for other type of problems - for sure conditional contextual bandits + maybe cats (which is right now is implemented as transformer to cb data). Maybe we could have package per problem type (only cb for now)?
- helpers / utils - looks like right now we have only ds_parse.py from this category. Probably it should be either completely removed or renamed to something from utils.py in future, but it is probably ok to leave it on higher level for now
- executable scripts - looks like it is basic_isage.py. If we think about estimators library as a library, this one is actually just example of possible usage, so it should not be part of the pip package and should stay either on level above or in separate examples/scripts folder.

**Review by ataymano — COMMENTED** [review]
Pseudo inverse is estimator for slates , not for conditional contextual bandits.

**Comment by JuiP** [comment]
The documentation [here](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Slates) suggested that slates builds upon ccb...

> Pseudo inverse is estimator for slates, not for conditional contextual bandits.

So then what would go under conditional contextual bandits? Also, do I rename the folder to slates?

**Comment by ataymano** [comment]
> The documentation [here](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Slates) suggested that slates builds upon ccb...
> 
> > Pseudo inverse is estimator for slates, not for conditional contextual bandits.
> 
> So then what would go under conditional contextual bandits? Also, do I rename the folder to slates?

Vw is built in the way that all problems are implemented as reductions to other (implemented before) problems.
So, conditional contextual bandits is implemented as reduction to contextual bandits and slates is implemented as reduction to conditional contextual bandits.
But from the interface point of view, these 3 problems are different:
1) Estimators for CB: taking (p_log:float, p_pred:float, r:float)
2) Estimators for CCB: taking (p_log:list[float], p_pred:list[float], r:list[float])
3) Estimators for slates: taking (p_log:list[float], p_pred:list[float], r:float)

Right now there is no CCB estimators implemented in this repo, so I suggest to move pseudo_inverse to slates folder, and conditional contextual bandits can be removed for now.

**Comment by ataymano** [comment]
Also, there is a convention in python package/module naming:
https://www.python.org/dev/peps/pep-0008/#package-and-module-names
which is violated if we name them ContextualBandits / ConditionalContextualBandits / Slates.
Can we name them either contextual_bandits / slates or cb / slates?

**Comment by JuiP** [comment]
> > The documentation [here](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Slates) suggested that slates builds upon ccb...
> > > Pseudo inverse is estimator for slates, not for conditional contextual bandits.
> > 
> > 
> > So then what would go under conditional contextual bandits? Also, do I rename the folder to slates?
> 
> Vw is built in the way that all problems are implemented as reductions to other (implemented before) problems.
> So, conditional contextual bandits is implemented as reduction to contextual bandits and slates is implemented as reduction to conditional contextual bandits.
> But from the interface point of view, these 3 problems are different:
> 
> 1. Estimators for CB: taking (p_log:float, p_pred:float, r:float)
> 2. Estimators for CCB: taking (p_log:list[float], p_pred:list[float], r:list[float])
> 3. Estimators for slates: taking (p_log:list[float], p_pred:list[float], r:float)
> 
> Right now there is no CCB estimators implemented in this repo, so I suggest to move pseudo_inverse to slates folder, and conditional contextual bandits can be removed for now.

Got it! Thank you so much for the clarification! :grin:

**Review by ataymano — APPROVED** [review]
Looks good to me. But let's rename PR to something like "Package file structure update" before merge - otherwise it looks a bit confusing

**Comment by JuiP** [comment]
@ataymano Agreed! I have changed the title of this PR

**Comment by ataymano** [comment]
Great, thank you! @cheng-tan, can you please take a look as well?

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import setuptools
+
+with open("README.md", "r") as f:
+    long_description = f.read()
+
+setuptools.setup(
+    name="vw-estimators",
+    version="0.0.1",
+    description="Python package of estimators to perform off-policy evaluation ",
```
Comment (cheng-tan) [inline setup.py:9]: remove the extra space at the end of description

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    url="https://github.com/VowpalWabbit/estimators.git",
+    license="BSD 3-Clause License",
+    classifiers=[
+        "Intended Audience :: Science/Research",
+        "License :: OSI Approved :: BSD License",
+        "Programming Language :: Python :: 3.6",
+        "Operating System :: OS Independent",
+        "Topic :: Scientific/Engineering"
+    ],
+    packages=["estimators"],
```
Comment (cheng-tan) [inline setup.py:21]: should we list all the packages containing __init__.py here? "estimators", "estimators.contextual_bandits", "estimators.slates"

Comment (JuiP) [reply setup.py]: Ah Yes! Thanks for noticing... I have added it in 61eb892. Please have a look...

**Pull request merged**